crates/sui-core/src/authority/authority_test_utils.rs

```diff
@@ -97,9 +99,33 @@ pub async fn execute_certificate_with_execution_error(
         state_acc.accumulate_cached_live_object_set_for_testing(include_wrapped_tombstone);
 
     if with_shared {
-        send_consensus(authority, &certificate).await;
-        if let Some(fullnode) = fullnode {
-            send_consensus(fullnode, &certificate).await;
+        if fake_consensus {
+            send_consensus(authority, &certificate).await;
+            if let Some(fullnode) = fullnode {
+                send_consensus(fullnode, &certificate).await;
+            }
+        } else {
+            // Just set object locks directly if send_consensus is not requested.
+            authority
+                .epoch_store_for_testing()
+                .assign_shared_object_versions_for_tests(
```

Comment:
How is this differently from the call to `send_consensus`, which will also eventually assign shared object versions? Is it mostly about whether to trigger any randomness related code?

Reply:
send_consensus runs the actual consensus handler code, which sees that DKG is not finished, and defers the tx without assigning versions. Then the tx runner test fails to execute the tx without versions assigned.

Reply:
`fake_consensus` is not very intuitive to me.
Also curious `send_consensus` also enqueues the transaction, do we not need to do that for the other case?
Overall I feel there is quite a diverge in these two paths with differences that is very hard to reason about, i.e. as a caller how to decide whether to pass fake_consensus to true or false. Could you work on it a bit more to see if we could make it clearer and easier to reason about as caller of this function?

Reply:
The enqueuing is extraneous because the very next thing transactional test runner does is call try_execute_immediately.

Can you explain more about what is unintuitive? `fake_consensus` means that if it's true, consensus behavior will be faked (by invoking the consensus handler), and if it's false, it won't be. It seems reasonable to allow tests to do either? In any case, the only two places it's invoked from are transactional test runner (for which we don't want fake consensus), and the few unit tests that call it via `send_and_confirm_transaction_` (for which I set it to `true` to keep behavior on those tests unchanged). It's not like this param is leaking out into many places.

Reply:
Well in either case it won't invoke consensus handler. In the send_consensus path, it just calls `process_consensus_transactions_for_tests` pretending the transactions come out of the handler. It is not obvious to me when one would chose over the other.

Reply:
Sorry, I'm not sure if you mean something different by consensus handler. What I meant is that it invokes `process_consensus_transactions_and_commit_boundary`, which is what I have been calling the consensus handler.

I agree it's not obvious in the abstract, but presumably if you're writing a unit test of certain behavior, you would know based on the thing you're testing whether it matters that the real consensus handling function is invoked, or if you just need versions to be assigned?

What would you like me to do here?

Reply:
It's less about what people are testing, but about what might not work under the testing setup.
To be more specific, before this change, we have been using `process_consensus_transactions_and_commit_boundary`, which isn't because we want to specifically test how consensus handling works but try to make sure that the code path in tests is as identical as possible to prod, so that it increases test coverage.
Now in this case, since the transactions actually use the randomness object, and hence leading to the call to `process_consensus_transactions_and_commit_boundary` break when there is no real consensus running, it won't work anymore, and hence we are switching to just assign versions.
So my question is:
1. Is using randomness the only thing that would cause the call to `process_consensus_transactions_and_commit_boundary` not work under tests? Could we document why?
2. If (1) is true, should we just check whether the transaction contains the randomness object, and if so call the right path, instead of asking the caller to decide the right parameter to pass.
3. If we are leaving the caller to decide the value of the parameter, it may be better to be more explicit, such as naming the parameter something like "uses_randomess" (but then 2 may be better here if randomness is the only thing).
4. Is there a way to continue using the same path?

Reply:
1. I don't know. This affects literally every test that uses a test authority, which is a lot of them. Hence why I didn't want to change behavior on any test where it wasn't necessary.
2. That seems pretty unexpected and "magical" to me
3. I think the parameter should be named after what it does. Why should something called "uses_randomness" affect the simulation mode for consensus?
4. This is a good idea. I tried it out, set `fake_consensus` to be true for everything, and the answer seems to be no. It broke some other simtests to not have `process_consensus_transactions_and_commit_boundary` be called

The change in this PR enables the transactional tests, which are totally artificial and don't need consensus, to behave differently from other tests, which are less artifical and do need consensus,.

I totally agree with you that the test infra overall could be further improved. But maybe it's worth going ahead to merge this change as an imperfect-but-positive step forward, to have randomness be added in to most of these tests (instead of disabled wholesale) and unblock you?

Reply:
Sounds good! Thank you for taking care of this!